cmd/fs-v1-multipart.go

```diff
@@ -345,126 +211,58 @@ func (fs fsObjects) ListMultipartUploads(bucket, object, keyMarker, uploadIDMark
-	uploadsPath := pathJoin(bucket, object, uploadsJSONFile)
-	rwlk, err := fs.rwPool.Create(pathJoin(fs.fsPath, minioMetaMultipartBucket, uploadsPath))
-	if err != nil {
-		return "", toObjectErr(errors.Trace(err), bucket, object)
-	}
-	defer rwlk.Close()
-
-	uploadIDPath := pathJoin(bucket, object, uploadID)
-	fsMetaPath := pathJoin(fs.fsPath, minioMetaMultipartBucket, uploadIDPath, fsMetaJSONFile)
-	metaFile, err := fs.rwPool.Create(fsMetaPath)
-	if err != nil {
-		return "", toObjectErr(errors.Trace(err), bucket, object)
-	}
-	defer metaFile.Close()
-
-	// Add a new upload id.
-	if err = fs.addUploadID(bucket, object, uploadID, initiated, rwlk); err != nil {
-		return "", toObjectErr(err, bucket, object)
-	}
-
-	// Write all the set metadata.
-	if _, err = fsMeta.WriteTo(metaFile); err != nil {
-		return "", toObjectErr(err, bucket, object)
-	}
-
-	// Return success.
-	return uploadID, nil
-}
-
 // NewMultipartUpload - initialize a new multipart upload, returns a
 // unique id. The unique id returned here is of UUID form, for each
 // subsequent request each UUID is unique.
 //
 // Implements S3 compatible initiate multipart API.
-func (fs fsObjects) NewMultipartUpload(bucket, object string, meta map[string]string) (string, error) {
+func (fs *fsObjects) NewMultipartUpload(bucket, object string, meta map[string]string) (string, error) {
 	if err := checkNewMultipartArgs(bucket, object, fs); err != nil {
-		return "", err
+		return "", toObjectErr(errors.Trace(err), bucket)
 	}
 
 	if _, err := fs.statBucketDir(bucket); err != nil {
-		return "", toObjectErr(err, bucket)
+		return "", toObjectErr(errors.Trace(err), bucket)
 	}
 
-	// Hold the lock so that two parallel complete-multipart-uploads
-	// do not leave a stale uploads.json behind.
-	objectMPartPathLock := fs.nsMutex.NewNSLock(minioMetaMultipartBucket, pathJoin(bucket, object))
-	if err := objectMPartPathLock.GetLock(globalOperationTimeout); err != nil {
-		return "", err
+	uploadID := mustGetUUID()
+	uploadIDDir := fs.getUploadIDDir(bucket, object, uploadID)
+
+	err := mkdirAll(uploadIDDir, 0755)
+	if err != nil {
+		return "", errors.Trace(err)
 	}
-	defer objectMPartPathLock.Unlock()
 
-	return fs.newMultipartUpload(bucket, object, meta)
-}
+	// Initialize fs.json values.
+	fsMeta := newFSMetaV1()
+	fsMeta.Meta = meta
 
-// Returns if a new part can be appended to fsAppendDataFile.
-func partToAppend(fsMeta fsMetaV1, fsAppendMeta fsMetaV1) (part objectPartInfo, appendNeeded bool) {
-	if len(fsMeta.Parts) == 0 {
-		return
+	fsMetaBytes, err := json.Marshal(fsMeta)
+	if err != nil {
+		return "", errors.Trace(err)
 	}
 
-	// As fsAppendMeta.Parts will be sorted len(fsAppendMeta.Parts) will naturally be the next part number
-	nextPartNum := len(fsAppendMeta.Parts) + 1
-	nextPartIndex := fsMeta.ObjectPartIndex(nextPartNum)
-	if nextPartIndex == -1 {
-		return
+	if err = ioutil.WriteFile(pathJoin(uploadIDDir, fsMetaJSONFile), fsMetaBytes, 0644); err != nil {
```

Comment:
You can use WriteTo interface here @krishnasrinivas and change the WriteTo to be.

```go
func (m *fsMetaV1) WriteTo(w io.Writer) (n int64, err error) {
        var metadataBytes []byte
        metadataBytes, err = json.Marshal(m)
        if err != nil {
                return 0, errors.Trace(err)
        }

        t, ok  := w.(*lock.LockedFile)
        if ok {
              if err =  t.Truncate(0); err != nil {
                return 0, errors.Trace(err)
              }
        }

        if _, err = lk.Write(metadataBytes); err != nil {
                return 0, errors.Trace(err)
        }

        // Success.
        return int64(len(metadataBytes)), nil
}
```

Reply:
I thought about this, but using WriteTo() does not give any advantage over ioutil.WriteFile().

On the other hand if WriteTo and ReadFrom takes lock.LockedFile the advantage is compiler always type checks that it reads and writes from a locked file. i.e we can have a convention that all access to the main namespace fs.json should be done through fsMeta.WriteTo and fsMeta.ReadFrom

Reply:
Okay that makes sense @krishnasrinivas